### ISSUE TITLE
Research base map alternatives

### ISSUE THREAD

**Issue by idfx** [AmericanRedCross/hacmap#10]
Research alternatives for base map solutions to use on the UI. Keep in mind offline work.

**Comment by danbjoseph** [comment]
tldr - 3D tiles are cool but probably too much effort for not much payoff for this particular project at this time
   
I was looking at 3D tiles, such as what is possible with [Cesium](https://cesium.com/cesiumjs/#exploring-openstreetmap-with-3d-tiles). The JavaScript client, as well as the tile builder and server, are all Apache License v2.0. It would be cool to be able to load our app with recent satellite imagery scenes (e.g. post-disaster), and drape it over an open source of digital elevation data such as the [NASA SRTM Plus](https://www2.jpl.nasa.gov/srtm/) (30 meter resolution) dataset. Cons: probably complicated?, maybe computationally expensive to process and serve the tiles?, and I’m not sure navigating a 3D map is particularly intuitive for most people (hard without a mouse and even then can be hard to get the zoom, pan, rotate, etc all figured out smoothly). 

<img width="1215" alt="Screen Shot 2021-02-22 at 8 43 19 AM" src="https://user-images.githubusercontent.com/4806884/108717702-91f7ba00-74eb-11eb-87c2-814722eaca43.png">

I think it may be already quite dated, but [this blog post](https://developer.here.com/blog/visualizing-large-scale-terrain-with-open-source-tool-tin-terrain) from Sep 2018 discusses using Tin-Terrain in combination with CesiumJS to privately or publicly host your own 3D terrain map. In addition to Cesium's offerings, there is [harp.gl](https://github.com/heremaps/harp.gl), an experimental and work in progress open-source 3D map rendering engine written in TypeScript. The *experimental and work in progress* makes me hesitant to consider this one too seriously at this point. 

Mapbox GL JS v2 is sick, with amazing 3D maps and features like a Sky API... _"As the Camera API increases map pitch and exposes the horizon, the Sky API simulates the sun’s position based on geographic location and time of day."_ But there are limitations to the license and I don't think we can use it for a self contained App.
<img width="989" alt="Screen Shot 2021-02-22 at 9 10 01 AM" src="https://user-images.githubusercontent.com/4806884/108719589-c10f2b00-74ed-11eb-9c3e-259779b35e19.png">

However, we might be able to offer some sense of terrain using elevation data and hill-shading techniques on a 2D map. Mapbox does this with their Streets tile style.
<img width="1325" alt="Screen Shot 2021-02-22 at 9 04 17 AM" src="https://user-images.githubusercontent.com/4806884/108719081-26165100-74ed-11eb-8e1e-8e617025cbfa.png">
Their controls allow you to tilt the map (and see that it's a flat surface despite the appearance of terrain).
<img width="946" alt="Screen Shot 2021-02-19 at 4 37 32 PM" src="https://user-images.githubusercontent.com/4806884/108719904-26fbb280-74ee-11eb-963f-fb44344260ad.png">
Limiting map navigation to left-right-up-down and in-out, should be easier for majority of users.

**Comment by danbjoseph** [comment]
For the client, I think there's 3 good options:
- [OpenLayers](https://openlayers.org/) makes it easy to put a dynamic map in any web page. It can display map tiles, vector data and markers loaded from any source. (2-clause BSD License - also known as the FreeBSD)
- [tangram](https://github.com/tangrams/tangram) is a JavaScript library for rendering 2D & 3D maps live in a web browser with WebGL (MIT license)
- [Leaflet](https://leafletjs.com/) is the leading open-source JavaScript library for mobile-friendly interactive maps, however it does require the a use of some sort of [plugin](https://leafletjs.com/plugins.html#vector-tiles) to render vector tiles 
- [MapLibre GL](https://github.com/maplibre/maplibre-gl-js) a community led [fork](https://www.maptiler.com/news/2021/01/mapbox-gl-open-source-fork/) derived from mapbox-gl-js prior to their switch to a non-OSS license
- this is an interesting [comparison of vector tile rendering](https://map-compare.apps.zerospatial.com/)

For the Server. I think the layer of app data (e.g. image points, extracted road lines) needs to be served as vector tiles for interaction. However, the base map could be either vector or raster. I've found 3 options that seem promising:
- [t-rex](https://github.com/t-rex-tileserver/t-rex/#t-rex) is a vector tile server specialized on publishing MVT tiles from your own data (MIT license). I'm not sure 
- [tegola](https://github.com/go-spatial/tegola) is a vector tile server delivering Mapbox Vector Tiles with support for PostGIS and GeoPackage data providers
- Apache + Mapnik + PostgreSQL as described on [Switch2OSM](https://switch2osm.org/serving-tiles/) for raster tiles
- [OpenMapTiles](https://openmaptiles.org/) seems to provide all the pieces for data import through to serving the tiles and also has stylesheets for the client rendering, but the documentation seems a little sparse and hard to follow at times (they'd rather you use their MapTiler service for your needs) 

Tool data (image points, road lines, etc):
- explore [serving dynamic vector tiles from PostGIS](https://info.crunchydata.com/blog/dynamic-vector-tiles-from-postgis)
- also, [pg_tileserv](https://github.com/CrunchyData/pg_tileserv) is a fast PostGIS-only tile server in Go

If we wanted a way to allow the user to load a different style sheet, we would probably want vector tiles for the basemap. Two OSS style editors (both mentioned in the t-rex project) are below. Could the app allow the user to change the map base style by uploading a new stylesheet?
- tangram-play ([repo](https://github.com/tangrams/tangram-play) & [demo](https://play.tangram.city/#15.6583/5.0069/-9.0363))
- [maputnik](https://maputnik.github.io/)

Base map data.
- [osm2pgsql](https://osm2pgsql.org/) for importing an .osm.pbf file into a PostGIS database
- how to have a simplified low detail world map, and allow a user to load/update areas of interest for a more detailed map of, for example, a particular country?
- blog post tutorial from Jan 2020 on [loading OSM data to PostGIS](https://blog.rustprooflabs.com/2020/01/postgis-osm-load-2020)
- [Creating tiles using Mapnik and generate_tiles.py](https://wiki.openstreetmap.org/wiki/Creating_your_own_tiles#Creating_tiles_using_Mapnik_and_generate_tiles.py)

**Comment by danbjoseph** [comment]
I think we should use PostgreSQL + [PostGIS](https://postgis.net/) for the data storage. So my main questions are:

- Deciding on the vector tile server to sit atop the PostGIS server. There’s no editing of the OSM data (so I don’t think I need osm-seed). The point and line data from the image processing is added to over time when more images are uploaded, but it’s not edited by the user through the map.
   - I was looking at [Tegola](https://github.com/go-spatial/tegola)
   - Based on this [Crunchy Data blog post](https://blog.crunchydata.com/dynamic-vector-tiles-from-postgis), also looking at [Martin](https://github.com/urbica/martin)
   - And also was considering [t-rex](https://github.com/t-rex-tileserver/t-rex/)
- Getting an OSM extract..
   - Geofabrik provides downloads but not every country is available. I guess I could download at subregion level (e.g. South America) and then use osmium-extract to get just a smaller extent?
   - https://export.hotosm.org/en/v3/ provides custom extents but does checking all the categories in the data configuration tab mean all OSM data is included?
- Do I need other map data, especially if I want a simple global overview, for example Natural Earth land/ocean/boundaries?
- The process for ingesting OSM data (I don’t have to update from changeset data or anything, but I want to be able to ingest a new extract at a later date to overwrite and update the basemap if needed)..
   - Reading the osm2pgsql documentation I see the available outputs are: flex, pgsql (default), gazetteer, multi, and null. Of which “flex” requires a Lua file containing configuration settings. Should I use osm2pgsql for the ingest and then is there a source for Lua configuration files that align with also available stylesheets for one of the client-side map renderers?
- Picking a client-side map renderer and figuring out the process for getting the basemap styled (hopefully leveraging existing cartographic work so we don’t have to define the map styling completely from scratch)..
   -  I don’t want to use Mapbox GL v2 because it charges for map loads even if using your own tiles – what about MapLibre GL? Leaflet would be with the Leaflet.VectorGrid plugin?

**Comment by danbjoseph** [comment]
- PostgreSQL/[PostGIS](https://postgis.net/) for the data storage
- [Tegola](https://github.com/go-spatial/tegola) to serve vector tiles
- [Open Layers](https://github.com/openlayers/openlayers#openlayers) to render the maps in the browser

**Comment by idfx** [comment]
Assigning to @rahulthewall for visibility

**Comment by idfx** [comment]
The idea @rahulthewall is that now you can review the alternatives that @danbjoseph outlined here and start a conversation leading to a decision. Sounds good?
